inc/integrations/class-coral.php

```diff
@@ -117,40 +150,109 @@ public function get_endpoint_settings(): array {
 	 *
 	 * @param \WP_REST_Request $request The request object.
 	 */
-	public function process_endpoint_request( \WP_REST_Request $request ) {
+	public function process_validate_endpoint_request( \WP_REST_Request $request ) {
 		// Allow access from the frontend.
 		header( 'Access-Control-Allow-Origin: ' . home_url() );
 
-		$user = sanitize_text_field( $request->get_param( 'user' ) );
-
 		$user_obj = [
-			'id'       => '628bdc61-6616-4add-bfec-dd79156715d4', // The ID should come from the Pico verification payload.
-			'email'    => $user,
-			'username' => explode( '@', $user )[0], // The username should come from the Pico verification payload.
+			'id'    => sanitize_text_field( $request->get_param( 'id' ) ),
+			'email' => sanitize_text_field( $request->get_param( 'email' ) ),
 		];
 
 		// Verify the user's credentials.
-		$verified_user = apply_filters( 'wp_irving_verify_coral_user', $user_obj );
+		$verified_data = apply_filters( 'wp_irving_verify_coral_user', $user_obj );
 
 		// Bail early if the verified user doesn't exist.
-		if ( empty( $verified_user ) ) {
+		if ( empty( $verified_data ) || ! $verified_data ) {
 			return [ 'status' => 'failed' ];
 		}
 
-		$credentials = [
-			'jti'  => uniqid(),
-			'exp'  => time() + ( 90 * DAY_IN_SECONDS ), // JWT will expire in 90 days.
-			'iat'  => time(),
-			'user' => [
-				'id'       => $verified_user['id'],
-				'email'    => $verified_user['email'],
-				'username' => $verified_user['username'],
-			],
+		// Check for existing username, since the verification call returned 200 OK.
+		$username = $this->get_username( $verified_data['user']['email'] );
+
+		if ( '' !== $username ) {
+			$credentials = [
+				'jti'  => uniqid(),
+				'exp'  => time() + ( 90 * DAY_IN_SECONDS ), // JWT will expire in 90 days.
+				'iat'  => time(),
+				'user' => [
+					'id'       => $verified_data['id'],
+					'email'    => $verified_data['user']['email'],
+					'username' => $username,
+				],
+			];
+
+			return [
+				'status'           => 'success',
+				'require_username' => false,
+				'jwt'              => $this->build_jwt( $credentials ),
+			];
+		}
+
+		// If username is not set then,
+		return [
+			'status'            => 'success',
+			'require_username'  => true,
+			'username_set_hash' => $this->create_username_set_hash( $verified_data['id'], $verified_data['user']['email'] ),
 		];
+	}
+
+	/**
+	 * Accept requests to set a username for a given email address.
+	 *
+	 * @param \WP_REST_Request $request The request object.
+	 * @return array|\WP_REST_Response
+	 */
+	public function process_set_username_endpoint_request( \WP_REST_Request $request ) {
+		// Allow access from the frontend.
+		header( 'Access-Control-Allow-Origin: ' . home_url() );
```

Comment:
I believe the REST API adds this by default and should be unnecessary. If it _is_ needed, than we should probably be filtering the `http_origin` filter and let the `rest_send_cors_headers()` function handle this.